motr/io_pargrp.c

```diff
@@ -1868,8 +1868,9 @@ static int pargrp_iomap_dgmode_postprocess(struct pargrp_iomap *map)
 				 * limitation of file size.
 				 */
 				if (map->pi_state == PI_DEGRADED) {
-					map->pi_databufs[row][col] =
-					    data_buf_alloc_init(obj, PA_NONE);
+					if (map->pi_databufs[row][col] == NULL)
```

Comment:
Do you like to:
M0_ASSERT(map->pi_databufs[row][col] == NULL);

Reply:
During my debug, I observed that, sometime map->pi_databufs[row][col] points to some memory address, still we were going ahead and allocate more memory which lead to some memory leaks. (leaving some memory blocks un-referenced)
Hence I added check ,if it is NULL then only allocate buffer else use the existing memory blocks allocated.

Reply:
If map->pi_databufs[row][col] is not NULL, it must already have been allocated or must point to some existing memory.

So, I am not sure if this will lead to some problem, or some errors lead to this.

Reply:
I agree, It might be allocated earlier or pointing to some memory (Could be used or unused) 
  But before this fix too it was anyway getting overwritten irrespective of  any check, I just added check to avoid re- allocation, which eventually leads to definitely leak. 
  
Bug still could be there which is not yet uncover. 
I can revert this change till we investigate more why it is not freed or is it safe to use pre-allocated buffer here.

Reply:
@huanghua78  : it could be possible that, It has allocated some buffers during pargrp_iomap_init, when client know that there is a device failure instead of reusing the same buffer, it is allocating new buffers which is adding a leak.

Reply:
The buffers of the data units may already exist, because we re-read non-failed data units on degraded read. Read more about this here - https://github.com/Seagate/cortx-motr/issues/1817.

Reply:
Thanks @andriytk for clarification. 
If I understood correctly, these buffer are expected to be allocated during normal read and it is safe to use for re-read of some data units during degraded read, in that case current fix from this PR(allocate buffer only if map->pi_databufs are NULL/not allocated) is right. 

But If it is expected that buffers should be always pre-allocated at this point, then better to add `M0_ASSERT(map->pi_databufs[row][col] != NULL); `   and avoid/remove allocation of databufs here.  

 @andriytk  @huanghua78  Let me know your suggestion here.

Reply:
Your change looks fine to me.